pkg/kubefnord/unjoin.go

```diff
@@ -0,0 +1,285 @@
+	host, unjoiningClusterName string, removeFromRegistry, dryRun bool) error {
+
+	clusterClientset, err := util.ClusterClientset(clusterConfig)
+	if err != nil {
+		glog.V(2).Infof("Failed to get joining cluster clientset: %v", err)
+		return err
+	}
+
+	fedClientset, err := util.FedClientset(hostConfig)
+	if err != nil {
+		glog.V(2).Infof("Failed to get federation clientset: %v", err)
+		return err
+	}
+
+	glog.V(2).Infof("Performing preflight checks for unjoin cluster.")
+	err = performPreflightChecksForUnjoin(clusterClientset, unjoiningClusterName, host, federationNamespace)
+	if err != nil {
+		return err
+	}
+
+	if removeFromRegistry {
+		err = removeFromClusterRegistry(hostConfig, clusterConfig.Host, unjoiningClusterName, dryRun)
+		if err != nil {
+			return err
+		}
+	}
+
+	glog.V(2).Info("Deleting federated cluster resource")
+
+	_, err = deleteFederatedCluster(clusterClientset, fedClientset, federationNamespace,
+		unjoiningClusterName, dryRun)
+	if err != nil {
+		glog.V(2).Infof("Failed to delete federated cluster resource: %v", err)
+		return err
+	}
+
+	glog.V(2).Info("Deleted federated cluster resource")
+
+	glog.V(2).Infof("Deleting namespace %s in unjoining cluster", federationNamespace)
+
+	err = clusterClientset.CoreV1().Namespaces().Delete(federationNamespace,
+		&metav1.DeleteOptions{})
+	if err != nil {
+		glog.V(2).Infof("Could not delete %s namespace: %v", federationNamespace, err)
+		return err
+	}
+
+	glog.V(2).Infof("Deleted namespace %s in unjoining cluster", federationNamespace)
+
+	return nil
+}
+
+// performPreflightChecks checks that the host and unjoining clusters are in
+// a consistent state.
+func performPreflightChecksForUnjoin(clusterClientset client.Interface, name, host,
+	federationNamespace string) error {
+	// Make sure there is a existing service account in the unjoining cluster.
+	saName := util.ClusterServiceAccountName(name, host)
+	sa, err := clusterClientset.CoreV1().ServiceAccounts(federationNamespace).Get(saName,
+		metav1.GetOptions{})
+
+	if errors.IsNotFound(err) {
+		return fmt.Errorf("service account does not exist in unjoining cluster")
+	} else if err != nil {
+		return err
+	} else if sa != nil {
+		return nil
+	}
+
+	return nil
+}
+
+// removeFromClusterRegistry handles removing the cluster from the cluster registry and
+// reports progress.
+func removeFromClusterRegistry(hostConfig *rest.Config, host, unjoiningClusterName string,
+	dryRun bool) error {
+	// Get the cluster registry clientset using the host cluster config.
+	crClientset, err := util.ClusterRegistryClientset(hostConfig)
+	if err != nil {
+		glog.V(2).Infof("Failed to get cluster registry clientset: %v", err)
+		return err
+	}
+
+	glog.V(2).Info("Removing cluster from the cluster registry.")
+
+	err = unRegisterCluster(crClientset, host, unjoiningClusterName, dryRun)
+	if err != nil {
+		glog.V(2).Infof("Could not unregister cluster with the cluster registry: %v", err)
```

Comment:
"Could not remove cluster from the cluster registry"

Reply:
Done